src/validator.ts

```diff
@@ -188,49 +189,268 @@ export const possiblyReturningBlock = warning<Method>(node => {
+  const parents = targetSupertypes(node)
+  const hasSuperclass = notEmpty(parents.filter(_ => !!_ && _.is('Class')))
+  const lastParentInHierarchy = last(parents)
+  return !hasSuperclass || !!lastParentInHierarchy && lastParentInHierarchy.is('Class')
+})
+
+export const shouldMatchSuperclassReturnValue = error<Method>(node => {
+  if (!node.isOverride) return true
+  const overridenMethod = superclassMethod(node)
+  if (!overridenMethod || overridenMethod.isAbstract() || overridenMethod.isNative()) return true
+  const lastSentence = last(node.sentences())
+  const superclassSentence = last(overridenMethod.sentences())
+  return !lastSentence || !superclassSentence || lastSentence.is('Return') === superclassSentence.is('Return') || lastSentence.is('Throw') || superclassSentence.is('Throw')
+})
+
+export const shouldReturnAValueOnAllFlows = error<If>(node => {
+  const lastThenSentence = last(node.thenBody.sentences)
+  const lastElseSentence = last(node.elseBody.sentences)
+  // TODO: For Send, consider if expression returns a value
+  const singleFlow = !lastElseSentence && lastThenSentence && finishesFlow(lastThenSentence, node)
+
+  // Try expression is still pending
+  const rightCombinations: Record<string, string[]> = {
+    'Assignment': ['Assignment', 'Send', 'Throw'],
+    'Literal': ['Literal', 'New', 'Self', 'Send', 'Reference', 'Super', 'Throw'],
+    'New': ['Literal', 'New', 'Self', 'Send', 'Reference', 'Super', 'Throw'],
+    'Reference': ['Literal', 'New', 'Self', 'Send', 'Reference', 'Super', 'Throw'],
+    'Return': ['Return', 'Throw'],
+    'Self': ['Literal', 'New', 'Self', 'Send', 'Reference', 'Super', 'Throw'],
+    'Send': ['Literal', 'New', 'Return', 'Self', 'Send', 'Reference', 'Super', 'Throw'],
+    'Throw': ['Literal', 'New', 'Return', 'Self', 'Send', 'Reference', 'Super', 'Throw'],
+  }
+
+  const twoFlows = !!lastThenSentence && !!lastElseSentence && (rightCombinations[lastThenSentence.kind]?.includes(lastElseSentence.kind) || rightCombinations[lastElseSentence.kind]?.includes(lastThenSentence.kind))
+  const ifFlows = !!lastThenSentence && !!lastElseSentence && (lastThenSentence.is('If') || lastElseSentence.is('If'))
+  return singleFlow || twoFlows || ifFlows
+})
+
+export const shouldNotDuplicateFields = error<Field>(node =>
+  node.parent().allFields().filter(_ => _.name == node.name).length === 1
+)
+
+export const parameterShouldNotDuplicateExistingVariable = error<Parameter>(node => {
+  const nodeMethod = getVariableContainer(node)
+  const parameterNotDuplicated = (nodeMethod as Method).parameters?.filter(parameter => parameter.name == node.name).length <= 1
+  return parameterNotDuplicated && !hasDuplicatedVariable(nodeMethod, node.name)
+})
+
+export const shouldNotDuplicateLocalVariables = error<Variable>(node => {
+  if (node.parent().is('Program') || node.parent().parent().is('Program') || isGlobal(node)) return true
```

Comment:
Esta linea no la entendí, porqué el padre del padre is program?

Reply:
Buena pregunta, cuando le saco esa condición me está tirando

```bash
  1) Wollok Validations
       shouldNotDefineGlobalMutableVariables:
     Error: Missing parent in cache for node 478348b3-6cbf-4640-883c-5a7ad2c30f1d
      at Environment.parent (src/model.ts:167:11)
      at Environment.descriptor.value (src/model.ts:72:35)
      at getVariableContainer (src/validator.ts:393:35)
      at /home/dodain/workspace/wollok-dev/wollok-ts/src/validator.ts:289:21
      at /home/dodain/workspace/wollok-dev/wollok-ts/src/validator.ts:76:6
      at /home/dodain/workspace/wollok-dev/wollok-ts/src/validator.ts:465:36
      at Array.map (<anonymous>)
      at /home/dodain/workspace/wollok-dev/wollok-ts/src/validator.ts:465:8
      at applyReduce (src/model.ts:222:10)
      at /home/dodain/workspace/wollok-dev/wollok-ts/src/model.ts:221:16
```

por ahí me podés dar una mano para ver qué era lo que no guarda en la cache...

Reply:
Update: ahora entendí bien, los casos en los que entra son porque la variable está definida en un body que a su vez está dentro de un program:

- Variable variableInsideProgramAllowed Body Program
- Variable constantReferenceInsideProgramIsAlsoOk Body Program
- Variable someVariable Body Program
- Variable a Body Program
- Variable collection Body Program
- Variable anObject Body Program

Reply:
ahí encontré una mejor opción, que el container principal sea un Program